File: scrape.rb
require "nokogiri"
require "httparty"
require "tty-table"
require "launchy"
require "tty-prompt"

# Defining data structure to store scraped data
Book = Struct.new(:number, :url, :title, :author, :price)

# Initializing Prompt
@prompt = TTY::Prompt.new

# Defining scrape function
def scrape_list
  # Initializing the list of objects that will contain the scraped data
  @books = []

  # Current iteration
  i = 1

  # Max pages to scrape
  limit = 5

  # Iterate over the 5 pages
  while i <= limit
    # Initializing the page to scrape
    if @input == "All"
      @page_to_scrape = "https://www.barnesandnoble.com/b/books/_/N-1fZ29Z8q8"
      @label = " "
    elsif @input == "Teens&YA"
      @page_to_scrape = "https://www.barnesandnoble.com/b/books/teens-ya/_/N-1fZ29Z8q8Z19r4"
      @label = " Teens & YA "
    elsif @input == "Kids"
      @page_to_scrape = "https://www.barnesandnoble.com/b/books/kids/_/N-1fZ29Z8q8Ztu1"
      @label = " Kids "
    elsif @input == "Fiction"
      @page_to_scrape = "https://www.barnesandnoble.com/b/fiction/books/_/N-1fZ2usxZ29Z8q8"
      @label = " Fiction "
    elsif @input == "NonFiction"
      @page_to_scrape = "https://www.barnesandnoble.com/b/nonfiction/books/_/N-1fZ2urcZ29Z8q8"
      @label = " NonFiction "
    end
    if i != 1
      @page_to_scrape = @page_to_scrape + "?Nrpp=20&page=" + i.to_s
    end

    # Downloading the target web page
    response = HTTParty.get(@page_to_scrape)

    # Parsing the HTML document returned by the server
    document = Nokogiri::HTML(response.body)

    # Selecting all HTML product elements
    html_products = document.css("li.pb-s")

    index = 1 + ((i - 1) * 20)
    # Iterating over the list of HTML products
    html_products.each do |html_product|
      # Extracting the data of interest

      # URL of Book
      url = html_product.css("a").first.attribute("href").value
      # Name of Book
      title = html_product.css("h3.product-info-title a").first.text
      # Author of Book
      author = html_product.css("div.product-shelf-author a").first.text
      # Price of Book
      price = html_product.css("span.current a").first.text

      # Storing the scraped data in a Book object
      book = Book.new(index, url, title, author, price)

      # Adding the book to the list of books
      @books.push(book)

      index += 1
    end
    i += 1
  end
  puts "Here are the top 100 Best Selling#{@label}Books: "
end

def render_best_selling_books
  header_array = ["Number", "Title", "Author", "Price"]
  table_array = []
  @books.each do |item|
    temp_array = []
    temp_array << item[0]
    temp_array << item[2]
    temp_array << item[3]
    temp_array << item[4]
    table_array << temp_array
  end

  table = TTY::Table.new(header_array, table_array)

  puts table.render(:ascii)
end

def render_book_info
  header_array = ["Number", "Title", "Author", "Price"]
  table_array = []
  temp_array = []
  temp_array << @selected_book[0]
  temp_array << @selected_book[2]
  temp_array << @selected_book[3]
  temp_array << @selected_book[4]
  table_array << temp_array

  table = TTY::Table.new(header_array, table_array)

  puts table.render(:ascii)

  @input = @prompt.select("Would you like to open the Barnes & Noble page for this book?", %w(Yes No))

  if @input == "Yes"
    Launchy.open("https://www.barnesandnoble.com#{@selected_book[1]}")
  end
end

# START OF APP

run_app = true

puts "Welcome to the Best Selling Books Finder!"

@input = @prompt.select("Which best selling book list would you like to see today?", %w(All Teens&YA Kids Fiction NonFiction))

while run_app
  scrape_list()

  render_best_selling_books()

  @input = @prompt.ask("Please input a book number that you would like more information on or enter '0' to quit.") do |q|
    q.in "0-100"
    q.messages[:range?] = "%{value} out of expected range %{in}"
  end

  while @input != "0"
    @selected_book = @books[@input.to_i - 1]
    render_book_info()

    @input = @prompt.ask("If there is another book you would like more info, please enter that number or enter '0' to quit.") do |q|
      q.in "0-100"
      q.messages[:range?] = "%{value} out of expected range %{in}"
    end
  end
  if @input == "0"
    run_app = false
  end
end

File: scrape.js
const puppeteer = require("puppeteer");
const prompt = require("prompt-sync")();
const open = require("open");

class Book {
  constructor(rank, title, url, author, price, rating) {
    this.rank = rank;
    this.title = title;
    this.url = url;
    this.author = author;
    this.price = price;
    this.rating = rating;
  }
}

var books = [];

function scrapeList() {
  // START OF APP

  let runApp = true;

  console.log("Welcome to the Best-Selling Books Finder!");

  let input = prompt(
    "Which best-selling book list would you like to see today? Please enter: 'All', 'Teens&YA', 'Kids', 'Fiction', or 'NonFiction': "
  );

  let selectedURL = "";
  let label = "";

  if (input.toLowerCase() === "all") {
    selectedURL = "https://www.barnesandnoble.com/b/books/_/N-1fZ29Z8q8";
    label = " ";
  } else if (input.toLowerCase() === "teens&ya") {
    selectedURL =
      "https://www.barnesandnoble.com/b/books/teens-ya/_/N-1fZ29Z8q8Z19r4";
    label = " Teens&YA ";
  } else if (input.toLowerCase() === "kids") {
    selectedURL =
      "https://www.barnesandnoble.com/b/books/kids/_/N-1fZ29Z8q8Ztu1";
    label = " Kids ";
  } else if (input.toLowerCase() === "fiction") {
    selectedURL =
      "https://www.barnesandnoble.com/b/fiction/books/_/N-1fZ2usxZ29Z8q8";
    label = " Fiction ";
  } else if (input.toLowerCase() === "nonfiction") {
    selectedURL =
      "https://www.barnesandnoble.com/b/nonfiction/books/_/N-1fZ2urcZ29Z8q8";
    label = " NonFiction ";
  }

  (async function scrape() {
    const browser = await puppeteer.launch({ headless: true });
    try {
      // scraping logic comes here...
      const page = await browser.newPage();
      await page.setUserAgent(
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4298.0 Safari/537.36"
      );

      await page.goto(selectedURL);

      await page.waitForSelector("ol");

      // Title of Book
      const titles = await page.$$eval("h3.product-info-title a", (nodes) =>
        nodes.map((n) => n.innerText)
      );

      // URL of Book
      const urls = await page.$$eval("h3.product-info-title a", (nodes) =>
        nodes.map((n) => n.href)
      );

      // Author of Book
      const authors = await page.$$eval("div.product-shelf-author", (nodes) =>
        nodes.slice(0, 20).map((n) => n.querySelector("a").innerText)
      );

      // Price of Book
      const prices = await page.$$eval("span.current a", (nodes) =>
        nodes.map((n) => n.innerText)
      );

      // Rating of Book
      const ratings = await page.$$eval("div.bv-off-screen", (nodes) =>
        nodes.map((n) => n.innerText)
      );

      await browser.close();

      for (let i = 0; i < titles.length; i++) {
        let book = new Book(
          i + 1,
          titles[i],
          urls[i],
          authors[i],
          prices[i],
          ratings[i]
        );
        books.push(book);
      }

      // Using the rank as the table id instead of the default id
      const transformed = books.reduce((acc, { rank, ...x }) => {
        acc[rank] = x;
        return acc;
      }, {});

      console.log(`Here are the top 20 Best-Selling${label}Books: `);
      console.table(transformed, ["title", "author", "price", "rating"]);

      while (runApp) {
        input = prompt(
          "Please input a book number that you would like more information on or enter 'quit' to quit. "
        );

        while (input.toLowerCase() !== "quit") {
          let selectedBook = transformed[parseInt(input)];

          console.table([selectedBook], ["title", "author", "price", "rating"]);

          // Scraping Logic for individual page
          await (async function scrape() {
            const browser = await puppeteer.launch({ headless: true });
            try {
              // scraping logic comes here...
              const page = await browser.newPage();
              await page.setUserAgent(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4298.0 Safari/537.36"
              );

              await page.goto(selectedBook["url"]);

              await page.waitForSelector("ol");

              // custReviewAuthor of Book
              const custReviewAuthor = await page.$$eval(
                "div.bv-author-avatar div.bv-content-author-name button.bv-author h3",
                (nodes) => nodes.map((n) => n.innerText)
              );

              // custReviewRating of Book
              const custReviewRating = await page.$$eval(
                "span.bv-rating-stars-container span.bv-off-screen",
                (nodes) => nodes.map((n) => n.innerText)
              );

              // custReviewTitle of Book
              const custReviewTitle = await page.$$eval(
                "div.bv-content-title-container h3.bv-content-title",
                (nodes) => nodes.map((n) => n.innerText)
              );

              // custReviewText of Book
              const custReviewText = await page.$$eval(
                "div.bv-content-summary-body-text p",
                (nodes) => nodes.map((n) => n.innerText)
              );

              await browser.close();

              if (custReviewAuthor.length === 0) {
                console.log(
                  "There appears to be no reviews for this book yet."
                );
              }

              for (let i = 0; i < custReviewAuthor.length; i++) {
                console.log(`Name: ${custReviewAuthor[i]}`);
                console.log(`Rating: ${custReviewRating[i]}`);
                console.log(`Review Title: ${custReviewTitle[i]}`);
                console.log(`Review: ${custReviewText[i]}`);
                console.log("--------------------------------");
              }
            } catch (e) {
              await browser.close();
              console.log("Error: ", e);
            }
          })();
          // End of Individual Scraping

          input = prompt(
            "Would you like to open the Barnes & Noble page for this book? Please enter: 'Yes' or 'No': "
          );

          if (input.toLowerCase() === "yes") {
            await open(selectedBook["url"]);
          }

          input = prompt(
            "If there is another book you would like more info, please enter that number or enter 'quit' to quit. "
          );
        }

        if (input === "quit") {
          runApp = false;
        }
      }
    } catch (e) {
      await browser.close();
      console.log("Error: ", e);
    }
  })();
}

scrapeList();
